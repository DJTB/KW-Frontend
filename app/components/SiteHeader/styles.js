import styled from 'styled-components';
import { transparentize } from 'polished';

import { centerByPadding, gutter } from 'shared/styles/layout';
import { white, whiteDark } from 'shared/styles/colors';
import { bottomLight } from 'shared/styles/shadows';

export const Header = styled.header`
  ${centerByPadding}
  ${gutter({ prop: 'margin', position: 'bottom', type: 'outer' })} /* logo unfortunately adds height already */
  position: relative;
  background-color: ${white};
  box-shadow: ${bottomLight};
  border-bottom: 1px solid ${transparentize(0.5, whiteDark)};
`;

export const Nav = styled.nav`
  ${gutter({ position: 'horizontal', type: 'outer' })} /* logo unfortunately adds height already */
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-end;
`;
