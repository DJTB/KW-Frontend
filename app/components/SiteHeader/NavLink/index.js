import React from 'react';
import PropTypes from 'prop-types';
import isNumber from 'lodash/isNumber';

import { Li, Link, LinkButton, Text, Count } from './styles';

NavLink.propTypes = {
  text: PropTypes.string.isRequired,
  count: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.number,
  ]),
  isOffCanvas: PropTypes.bool,
  handleLogout: PropTypes.func.isRequired,
};

NavLink.defaultProps = {
  count: false,
  isOffCanvas: false,
};

function NavLink({
  text,
  count,
  isOffCanvas,
  handleLogout,
}) {
  return (
    <Li isOffCanvas={isOffCanvas}>
      {text === 'logout' ? (
        <LinkButton onClick={handleLogout}>
          <Text>
            {text}
          </Text>
        </LinkButton>
      ) : (
        <Link
          plainLink
          to={`/${text}`}
        >
          <Text>
            {text}
            {isNumber(count) && <Count disabled={count < 1}>{count}</Count>}
          </Text>
        </Link>
        )
      }
    </Li>
  );
}

export default NavLink;
