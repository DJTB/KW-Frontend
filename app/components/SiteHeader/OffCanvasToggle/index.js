import React from 'react';
import PropTypes from 'prop-types';

import { Toggle, ToggleDot } from './styles';

OffCanvasToggle.propTypes = {
  isActive: PropTypes.bool,
  ariaControls: PropTypes.string.isRequired,
  handleToggle: PropTypes.func.isRequired,
};

OffCanvasToggle.defaultProps = {
  isActive: false,
};

function OffCanvasToggle({ isActive, handleToggle, ariaControls }) {
  return (
    <Toggle
      type="button"
      aria-label="navigation menu"
      aria-controls={ariaControls}
      aria-expanded={isActive}
      onClick={handleToggle}
    >
      <ToggleDot />
      <ToggleDot />
      <ToggleDot />
    </Toggle>
  );
}

export default OffCanvasToggle;
