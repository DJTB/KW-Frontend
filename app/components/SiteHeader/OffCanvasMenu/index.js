import React from 'react';
import PropTypes from 'prop-types';
import cuid from 'cuid';

import NavLink from '../NavLink';

import { Wrapper, Ul } from './styles';

OffCanvasMenu.propTypes = {
  id: PropTypes.string.isRequired,
  routes: PropTypes.array.isRequired,
  offsetTop: PropTypes.number.isRequired,
  isVisible: PropTypes.bool.isRequired,
};

function OffCanvasMenu({ id, routes, offsetTop, isVisible }) {
  return (
    <Wrapper
      id={id}
      offsetTop={offsetTop}
      isVisible={isVisible}
    >
      <Ul>
        {routes.map((route) => (
          <NavLink
            key={cuid()}
            isOffCanvas
            {...route}
          />
        ))}
      </Ul>
    </Wrapper>
  );
}

export default OffCanvasMenu;
