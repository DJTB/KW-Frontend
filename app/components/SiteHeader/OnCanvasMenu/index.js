import React from 'react';
import PropTypes from 'prop-types';
import cuid from 'cuid';

import NavLink from '../NavLink';
import { Ul } from './styles';

OnCanvasMenu.propTypes = {
  routes: PropTypes.array.isRequired,
};

function OnCanvasMenu({ routes }) {
  return (
    <Ul>
      {routes.map((route) => (
        <NavLink
          key={cuid()}
          {...route}
        />
      ))}
    </Ul>
  );
}

export default OnCanvasMenu;
