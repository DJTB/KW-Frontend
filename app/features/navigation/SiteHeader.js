import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { debounce, isEqual } from 'lodash';
import cuid from 'cuid';

import { breakpoints } from 'common/styles/media';
import { selectLocationPath } from 'common/selectors';

import Element from 'common/components/Element';
import LogoLink from 'common/components/LogoLink';
import Hamburger from './Hamburger';
import OffCanvasMenu from './OffCanvasMenu';

import { Header, Nav, NavLinks } from './styles';
import NavLink from './NavLink';

const PRIMARY_LINKS = [
  {
    name: 'lessons',
    route: '/lessons',
    hasCount: true,
  },
  {
    name: 'reviews',
    route: '/reviews',
    hasCount: true,
  },
];
const SECONDARY_LINKS = [
  { name: 'vocabulary', route: '/vocabulary' },
  { name: 'settings', route: '/settings' },
  { name: 'about', route: '/about' },
  { name: 'contact', route: '/contact' },
  { name: 'logout', route: '/logout' },
];

export class SiteHeader extends React.Component {
  static propTypes = {
    locationPath: PropTypes.string.isRequired,
  };

  state = {
    offCanvasMenuActive: false,
  };

  componentDidMount() {
    this.handleResize();
    window.addEventListener('resize', this.handleResize);
  }

  shouldComponentUpdate(nextProps, nextState) {
    const unchanged = isEqual(this.props, nextProps) && isEqual(this.state, nextState);
    return !unchanged;
  }

  componentDidUpdate(prevProps) {
    if (this.props.locationPath !== prevProps.locationPath) {
      this.hideOffCanvasMenu();
    }
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
    this.hideOffCanvasMenu();
  }

  handleResize = debounce(() => {
    this.hideOffCanvasMenu();
    this.setState(() => ({
      isWideViewport: window.innerWidth > breakpoints.md,
    }));
  }, 150);

  hideOffCanvasMenu = () => {
    this.setState({ offCanvasMenuActive: false });
    document.body.classList.remove('offCanvasMenu--isOpen');
  };

  showOffCanvasMenu = () => {
    this.setState({ offCanvasMenuActive: true });
    document.body.classList.add('offCanvasMenu--isOpen');
  };

  renderMenu = (links) => (
    <NavLinks>
      {links.map((link) => <NavLink key={cuid()} isOffCanvas={false} {...link} />)}
    </NavLinks>
  );

  render() {
    const { isWideViewport, offCanvasMenuActive } = this.state;
    const logoSize = isWideViewport ? '4.25rem' : '3.25rem';

    return isWideViewport === undefined ? null : (
      <Header
        innerRef={(node) => {
          this.HeaderRef = node;
        }}
      >
        <Nav>
          <LogoLink size={logoSize} />
          <Element flexRow flex="999 1 auto" justifyContent="space-between">
            {this.renderMenu(PRIMARY_LINKS)}
            {isWideViewport && this.renderMenu(SECONDARY_LINKS)}
          </Element>
          {!isWideViewport && <Hamburger onToggle={this.showOffCanvasMenu} />}
        </Nav>
        <OffCanvasMenu
          links={SECONDARY_LINKS}
          isVisible={offCanvasMenuActive}
          onClose={this.hideOffCanvasMenu}
        />
      </Header>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  locationPath: selectLocationPath,
});

export default connect(mapStateToProps)(SiteHeader);
