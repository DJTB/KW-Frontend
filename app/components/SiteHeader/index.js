import React from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';

import { breakpoints } from 'shared/styles/media';

import LogoLink from 'components/LogoLink';

import OnCanvasMenu from './OnCanvasMenu';
import OffCanvasToggle from './OffCanvasToggle';
import OffCanvasMenu from './OffCanvasMenu';
import { Header, Nav } from './styles';

// FIXME: selectors for counts
// the rest can probably just be local state, since we have a nav of specific routes only.
const devRoutes = [
  // {
  //   text: 'lessons',
  //   to: '/lessons',
  //   count: 0,
  //   isDisabled: false,
  // },
  {
    text: 'reviews',
    to: '/reviews',
    count: 0,
    isDisabled: false,
  },
  {
    text: 'vocabulary',
    to: '/vocabulary',
  },
  {
    text: 'settings',
    to: '/settings',
  },
  {
    text: 'about',
    to: '/about',
  },
  {
    text: 'contact',
    to: '/contact',
  },
  {
    text: 'logout',
    to: '/logout',
  },
];

class SiteHeader extends React.PureComponent {
  static propTypes = {
    routes: PropTypes.array.isRequired,
  }

  static defaultProps = {
    routes: devRoutes,
  }

  state = {
    headerHeight: 70, // ballpark fallback
    offCanvasMenuActive: false,
    offCanvasToggleVisible: false,
    onCanvasWidth: null,
    linkWidths: null,
  };

  componentDidMount() {
    this.handleResize();
    window.addEventListener('resize', this.handleResize);
    window.addEventListener('click', this.hideOffCanvasMenu);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
    window.removeEventListener('click', this.hideOffCanvasMenu);
  }

  handleResize = debounce(() => {
    this.setState(() => ({
      headerHeight: this.HeaderRef.clientHeight,
      offCanvasToggleVisible: window.innerWidth <= breakpoints.md,
    }));
  }, 150);

  hideOffCanvasMenu = () => {
    if (this.state.offCanvasMenuActive === true) {
      this.setState({ offCanvasMenuActive: false });
    }
  }

  handleToggleClick = (event) => {
    event.stopPropagation();
    this.setState(prevState => ({
      offCanvasMenuActive: !prevState.offCanvasMenuActive,
    }));
  }

  render() {
    let onCanvasRoutes = this.props.routes;
    let offCanvasRoutes = [];

    if (this.state.offCanvasToggleVisible) {
      onCanvasRoutes = this.props.routes.slice(0, 2);
      offCanvasRoutes = this.props.routes.slice(2);
    }

    return (
      <Header innerRef={(node) => { this.HeaderRef = node; }}>
        <Nav>
          <LogoLink />
          <OnCanvasMenu routes={onCanvasRoutes} />
          {this.state.offCanvasToggleVisible && (
            <OffCanvasToggle
              ariaControls="offCanvasMenu"
              isActive={this.state.offCanvasMenuActive}
              handleToggle={this.handleToggleClick}
            />
          )}
          <OffCanvasMenu
            id="offCanvasMenu"
            routes={offCanvasRoutes}
            offsetTop={this.state.headerHeight}
            isVisible={this.state.offCanvasMenuActive}
          />
        </Nav>
      </Header>
    );
  }
}

export default SiteHeader;
