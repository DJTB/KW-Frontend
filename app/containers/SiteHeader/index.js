import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import debounce from 'lodash/debounce';
import { clearToken } from 'utils/auth';
import { breakpoints } from 'shared/styles/media';

import LogoLink from 'components/LogoLink';
import { selectReviewCount } from 'containers/App/selectors';

import OnCanvasMenu from './OnCanvasMenu';
import OffCanvasToggle from './OffCanvasToggle';
import OffCanvasMenu from './OffCanvasMenu';
import { Header, Nav } from './styles';

class SiteHeader extends React.Component {
  static propTypes = {
    reviewCount: PropTypes.number,
  };

  static defaultProps = {
    reviewCount: 0,
  };

  state = {
    headerHeight: 70, // ballpark fallback
    offCanvasMenuActive: false,
    offCanvasToggleVisible: false,
  };

  componentDidMount() {
    this.handleResize();
    window.addEventListener('resize', this.handleResize);
    window.addEventListener('click', this.hideOffCanvasMenu);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
    window.removeEventListener('click', this.hideOffCanvasMenu);
  }

  handleResize = debounce(() => {
    this.setState(() => ({
      headerHeight: this.HeaderRef.clientHeight,
      offCanvasToggleVisible: window.innerWidth <= breakpoints.md,
    }));
  }, 150);

  hideOffCanvasMenu = () => {
    if (this.state.offCanvasMenuActive === true) {
      this.setState({ offCanvasMenuActive: false });
    }
  }

  handleToggle = (event) => {
    event.stopPropagation();
    this.setState((prevState) => ({
      offCanvasMenuActive: !prevState.offCanvasMenuActive,
    }));
  }

  handleLogout = () => {
    clearToken();
    this.props.history.push('/logout'); // eslint-disable-line react/prop-types
  }

  render() {
    const isWideViewport = !this.state.offCanvasToggleVisible;
    let onCanvasRoutes = [
      // { text: 'lessons', count: this.props.lessonCount },
      { text: 'reviews', count: this.props.reviewCount },
      { text: 'vocabulary' },
    ];

    let offCanvasRoutes = [
      { text: 'settings' },
      { text: 'about' },
      { text: 'contact' },
      { text: 'logout' },
    ];

    // show all routes in main menu if large screen
    if (isWideViewport) {
      onCanvasRoutes = onCanvasRoutes.concat(offCanvasRoutes);
      offCanvasRoutes = [];
    }

    return (
      <Header innerRef={(node) => { this.HeaderRef = node; }}>
        <Nav>
          <LogoLink />
          <OnCanvasMenu
            routes={onCanvasRoutes}
            handleLogout={this.handleLogout}
          />
          <OffCanvasToggle
            isVisible={this.state.offCanvasToggleVisible}
            isActive={this.state.offCanvasMenuActive}
            handleToggle={this.handleToggle}
            ariaControls="offCanvasMenu"
          />
          <OffCanvasMenu
            id="offCanvasMenu"
            routes={offCanvasRoutes}
            offsetTop={this.state.headerHeight}
            isVisible={this.state.offCanvasMenuActive}
            handleLogout={this.handleLogout}
          />
        </Nav>
      </Header>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  reviewCount: selectReviewCount,
});

export default connect(mapStateToProps)(SiteHeader);
