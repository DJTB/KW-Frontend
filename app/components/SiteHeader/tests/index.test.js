import React from 'react';
import { render } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import SiteHeader from '../index';

describe('<SiteHeader />', () => {
  it('should match baseline snapshot', () => {
    const renderedComponent = render(
      <MemoryRouter>
        <SiteHeader />
      </MemoryRouter>
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
