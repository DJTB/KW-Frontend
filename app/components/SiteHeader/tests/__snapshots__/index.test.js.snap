// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteHeader /> should match baseline snapshot 1`] = `
<header
  class="sc-gZMcBi frlyKr"
>
  <nav
    class="sc-gqjmRU mUhZH"
  >
    <a
      class="sc-bdVaJa bhRTrH"
      href="/"
      title="Return Home"
    >
      <image_mock
        class="sc-bwzfXH itbozR"
        height="100%"
        width="100%"
      />
    </a>
    <ul
      class="sc-bZQynM hRtPKZ"
    >
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/lesson"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            lessons
            <span
              class="sc-bxivhb iEfeYG"
            >
              0
            </span>
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/review"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            reviews
            <span
              class="sc-bxivhb iEfeYG"
            >
              0
            </span>
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/vocabulary"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            vocabulary
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/settings"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            settings
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/about"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            about
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/logout"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            logout
          </span>
        </a>
      </li>
    </ul>
    <div
      class="sc-dnqmqq gvSxfF"
      id="offCanvasMenu"
    >
      <ul
        class="sc-iwsKbI TzgrY"
      />
    </div>
  </nav>
</header>
`;

exports[`<SiteNav /> should match baseline snapshot 1`] = `
<header
  class="sc-gZMcBi frlyKr"
>
  <nav
    class="sc-gqjmRU mUhZH"
  >
    <a
      class="sc-bdVaJa bhRTrH"
      href="/"
      title="Return Home"
    >
      <image_mock
        class="sc-bwzfXH itbozR"
        height="100%"
        width="100%"
      />
    </a>
    <ul
      class="sc-bZQynM hRtPKZ"
    >
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/lesson"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            lessons
            <span
              class="sc-bxivhb iEfeYG"
            >
              0
            </span>
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/review"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            reviews
            <span
              class="sc-bxivhb iEfeYG"
            >
              0
            </span>
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/vocabulary"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            vocabulary
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/settings"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            settings
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/about"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            about
          </span>
        </a>
      </li>
      <li
        class="sc-EHOje ibQBqg"
      >
        <a
          class="sc-ifAKCX kGWvaB"
          href="/logout"
        >
          <span
            class="sc-htpNat eEEGMG"
          >
            logout
          </span>
        </a>
      </li>
    </ul>
    <div
      class="sc-dnqmqq gvSxfF"
      id="offCanvasMenu"
    >
      <ul
        class="sc-iwsKbI TzgrY"
      />
    </div>
  </nav>
</header>
`;
